File: wave_propagation_3D_obstacle.py
import os
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation
from mpl_toolkits.mplot3d import Axes3D

# Simulation parameters
Lx, Ly, Lz = 2, 1, 2         # Domain size
Nx, Ny, Nz = 150, 75, 150    # Grid points
dx, dy, dz = Lx / Nx, Ly / Ny, Lz / Nz
c = 150                      # Wave speed
dt = 0.00005                 # Time step
Nt = 800                     # Number of time steps

# CFL condition
if c * dt / min(dx, dy, dz) > 1:
    raise ValueError("CFL condition not satisfied. Adjust dx, dy, dz, or dt.")

# Initialize wave field
u = np.zeros((Nt, Nx, Ny, Nz))
x = np.linspace(0, Lx, Nx)
y = np.linspace(0, Ly, Ny)
z = np.linspace(0, Lz, Nz)
X, Y, Z = np.meshgrid(x, y, z, indexing='ij')

# Initial conditions: Gaussian pulse
u[0, :, :, :] = np.exp(-100 * ((X - 0.5)**2 + (Y - 0.5)**2 + (Z - 1)**2))
u[1, :, :, :] = u[0, :, :, :]  # Initial velocity = 0

# Defining a pillar in the region
pillar_start_x = int(0.7 * Nx)  # Start of the pillar in the X direction
pillar_end_x = int(0.85 * Nx)   # End of the pillar in the X direction
pillar_start_y = int(0.3 * Ny)  # Start of the pillar in the Y direction
pillar_end_y = int(0.7 * Ny)    # End of the pillar in the Y direction
pillar_start_z = 0              # Start of the pillar in the Z direction
pillar_end_z = int(0.6 * Nz)    # End of the pillar in the Z direction

# Set the wave field values to zero in the pillar region
u[:, pillar_start_x:pillar_end_x, pillar_start_y:pillar_end_y, pillar_start_z:pillar_end_z] = 0

# Finite difference method
for n in range(1, Nt - 1):
    for i in range(1, Nx - 1):
        for j in range(1, Ny - 1):
            for k in range(1, Nz - 1):
                u[n+1, i, j, k] = (2 * u[n, i, j, k] - u[n-1, i, j, k] +
                                   (c * dt / dx)**2 * (u[n, i+1, j, k] - 2 * u[n, i, j, k] + u[n, i-1, j, k]) +
                                   (c * dt / dy)**2 * (u[n, i, j+1, k] - 2 * u[n, i, j, k] + u[n, i, j-1, k]) +
                                   (c * dt / dz)**2 * (u[n, i, j, k+1] - 2 * u[n, i, j, k] + u[n, i, j, k-1]))
    # Ensure the pillar region remains zero
    u[n+1, pillar_start_x:pillar_end_x, pillar_start_y:pillar_end_y, pillar_start_z:pillar_end_z] = 0

# Directory to save frames
output_dir = "frames_wave_propagation_3D_obstacle"
os.makedirs(output_dir, exist_ok=True)

# Animation: Visualizing slices
fig = plt.figure(figsize=(19.2, 10.8), dpi=200)
ax = fig.add_subplot(111, projection='3d')
slice_idx = Nz // 2  # Middle slice in the z-direction
ax.view_init(elev=20, azim=80) # Set initial viewing angle

def update(frame):
    ax.clear()
    plot = ax.plot_surface(X[:, :, slice_idx], Y[:, :, slice_idx], u[frame, :, :, slice_idx], cmap="viridis")
    ax.set_title(f"3D Wave Propagation with an obstacle (Slice at z=1), Frame {frame}")
    ax.set_xlabel("x")
    ax.set_ylabel("y")
    ax.set_zlabel("u(x, y, t)")
    ax.set_zlim(-2, 2)
    ax.set_box_aspect([Lx, Ly, Lz])
    
    # Visualize the static pillar
    X_pillar, Y_pillar, Z_pillar = np.meshgrid(
        x[pillar_start_x:pillar_end_x],
        y[pillar_start_y:pillar_end_y],
        z[pillar_start_z:pillar_end_z],
        indexing='ij'
    )
    
    # Plot all six faces of the cuboid
    ax.plot_surface(X_pillar[:, :, 0], Y_pillar[:, :, 0], Z_pillar[:, :, 0], color='red', alpha=0.5)  # Bottom face
    ax.plot_surface(X_pillar[:, :, -1], Y_pillar[:, :, -1], Z_pillar[:, :, -1], color='red', alpha=0.5)  # Top face
    ax.plot_surface(X_pillar[:, 0, :], Y_pillar[:, 0, :], Z_pillar[:, 0, :], color='red', alpha=0.5)  # Front face
    ax.plot_surface(X_pillar[:, -1, :], Y_pillar[:, -1, :], Z_pillar[:, -1, :], color='red', alpha=0.5)  # Back face
    ax.plot_surface(X_pillar[0, :, :], Y_pillar[0, :, :], Z_pillar[0, :, :], color='red', alpha=0.5)  # Left face
    ax.plot_surface(X_pillar[-1, :, :], Y_pillar[-1, :, :], Z_pillar[-1, :, :], color='red', alpha=0.5)  # Right face
    
    # Save the current frame as a PNG image
    plt.savefig(f"{output_dir}/frame_{frame:04d}.png", dpi=200)
    return plot,''

# Create the animation
ani = FuncAnimation(fig, update, frames=Nt, interval=50)

# Show the plot (optional)
plt.show()

File: wave_propagation_3D_obstacle.cpp
#include <vector>
#include <cmath>
#include <iostream>
#include <GL/glut.h>
#include <GL/gl.h>
#include <chrono>
#include <thread>
#include <algorithm>

// Simulation parameters
const double Lx = 2.0, Ly = 1.0, Lz = 2.0;              // Domain size
const int Nx = 160, Ny = 80, Nz = 160;                  // Resolution - grid points
const double dx = Lx / Nx, dy = Ly / Ny, dz = Lz / Nz;
const double c = 1000.0;                                  // Wave speed
const double dt = 0.000005;                              // Time steps
const int Nt = 2000;                                     // Number of frames

// Pillar parameters
const int pillar_start_x = static_cast<int>(0.7 * Nx);
const int pillar_end_x = static_cast<int>(0.85 * Nx);
const int pillar_start_y = static_cast<int>(0.3 * Ny);
const int pillar_end_y = static_cast<int>(0.7 * Ny);
const int pillar_start_z = 0;
const int pillar_end_z = static_cast<int>(0.6 * Nz);

// Global variables
std::vector<std::vector<std::vector<double>>> u_current(Nx, std::vector<std::vector<double>>(Ny, std::vector<double>(Nz)));
std::vector<std::vector<std::vector<double>>> u_prev(Nx, std::vector<std::vector<double>>(Ny, std::vector<double>(Nz)));
std::vector<std::vector<std::vector<double>>> u_next(Nx, std::vector<std::vector<double>>(Ny, std::vector<double>(Nz)));
int currentStep = 0;
bool simulationRunning = true;

double calculateMaxStableC() {
    // CFL condition in 3D
    double denominator = dt * std::sqrt(1.0/(dx*dx) + 1.0/(dy*dy) + 1.0/(dz*dz));
    return 1.0 / denominator;
}

void init() {
    glEnable(GL_DEPTH_TEST);
    glEnable(GL_BLEND);
    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
    glClearColor(1.0, 1.0, 1.0, 1.0);
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    gluPerspective(45.0, 2.0, 0.1, 10.0);  // Modified aspect ratio for wider view
    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();
    
    // Adjusted camera position for better view of the elongated domain
    gluLookAt(1.0, 2.5, 1.25,    // Camera position: above and behind
              1.0, -1.0, 0.0,    // Look at point: center of x-axis
              0.0, 0.0, 1.0);    // Up vector: aligned with z-axis
}

void calculateNextStep() {
    if (currentStep >= Nt) {
        simulationRunning = false;
        return;
    }

    for (int i = 1; i < Nx - 1; i++) {
        for (int j = 1; j < Ny - 1; j++) {
            for (int k = 1; k < Nz - 1; k++) {
                // Skip calculation if point is inside pillar
                if (i >= pillar_start_x && i < pillar_end_x &&
                    j >= pillar_start_y && j < pillar_end_y &&
                    k >= pillar_start_z && k < pillar_end_z) {
                    u_next[i][j][k] = 0;
                    continue;
                }

                u_next[i][j][k] = 2.0 * u_current[i][j][k] - u_prev[i][j][k] +
                    (c * dt / dx) * (c * dt / dx) * (u_current[i+1][j][k] - 2.0 * u_current[i][j][k] + u_current[i-1][j][k]) +
                    (c * dt / dy) * (c * dt / dy) * (u_current[i][j+1][k] - 2.0 * u_current[i][j][k] + u_current[i][j-1][k]) +
                    (c * dt / dz) * (c * dt / dz) * (u_current[i][j][k+1] - 2.0 * u_current[i][j][k] + u_current[i][j][k-1]);
            }
        }
    }

    u_prev = u_current;
    u_current = u_next;
    currentStep++;
}

void drawPillar() {
    glColor4f(0.8f, 0.2f, 0.2f, 0.5f);  // Semi-transparent red
    
    float x1 = pillar_start_x * dx;
    float x2 = pillar_end_x * dx;
    float y1 = pillar_start_y * dy;
    float y2 = pillar_end_y * dy;
    float z1 = pillar_start_z * dz;
    float z2 = pillar_end_z * dz;

    // Draw the pillar as a solid cuboid
    glBegin(GL_QUADS);
    // Front face
    glVertex3f(x1, y1, z1);
    glVertex3f(x2, y1, z1);
    glVertex3f(x2, y1, z2);
    glVertex3f(x1, y1, z2);
    
    // Back face
    glVertex3f(x1, y2, z1);
    glVertex3f(x2, y2, z1);
    glVertex3f(x2, y2, z2);
    glVertex3f(x1, y2, z2);
    
    // Right face
    glVertex3f(x2, y1, z1);
    glVertex3f(x2, y2, z1);
    glVertex3f(x2, y2, z2);
    glVertex3f(x2, y1, z2);
    
    // Left face
    glVertex3f(x1, y1, z1);
    glVertex3f(x1, y2, z1);
    glVertex3f(x1, y2, z2);
    glVertex3f(x1, y1, z2);
    
    // Top face
    glVertex3f(x1, y1, z2);
    glVertex3f(x2, y1, z2);
    glVertex3f(x2, y2, z2);
    glVertex3f(x1, y2, z2);
    
    // Bottom face
    glVertex3f(x1, y1, z1);
    glVertex3f(x2, y1, z1);
    glVertex3f(x2, y2, z1);
    glVertex3f(x1, y2, z1);
    glEnd();
}

// Add these helper functions after the global variables:
struct Color {
    float r, g, b;
};

Color getHeightColor(float height) {
    // Normalize height to [-1,1] range and apply non-linear scaling
    float h = height;
    
    // Apply non-linear scaling to make middle range more sensitive
    if (std::abs(h) <= 0.4f) {
        // Scale middle range (-0.4 to 0.4) to full range
        h = h * 1.25f;
    } else {
        // Compress outer ranges
        h = (h > 0) ? 0.5f + (h - 0.4f) * 0.5f : -0.5f + (h + 0.4f) * 0.5f;
    }
    
    // Now normalize to [0,1] range
    h = (h + 1.0f) / 2.0f;
    
    // Define color gradient
    if (h < 0.2f) {
        // Deep blue to blue
        return {0.0f, 0.0f, 0.3f + 3.5f * h};
    } else if (h < 0.4f) {
        // Blue to cyan
        float t = (h - 0.2f) * 5.0f;
        return {0.0f, t, 1.0f};
    } else if (h < 0.6f) {
        // Cyan to green
        float t = (h - 0.4f) * 5.0f;
        return {0.0f, 1.0f, 1.0f - t};
    } else if (h < 0.8f) {
        // Green to yellow
        float t = (h - 0.6f) * 5.0f;
        return {t, 1.0f, 0.0f};
    } else {
        // Yellow to red
        float t = (h - 0.8f) * 5.0f;
        return {1.0f, 1.0f - t, 0.0f};
    }
}

void display() {
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
    
    // Draw coordinate axes
    glBegin(GL_LINES);
    glColor3f(0.3, 0.3, 0.3);
    glVertex3f(0.0, 0.0, 0.0);
    glVertex3f(2.0, 0.0, 0.0);
    glVertex3f(0.0, 0.0, 0.0);
    glVertex3f(0.0, 1.0, 0.0);
    glVertex3f(0.0, 0.0, 0.0);
    glVertex3f(0.0, 0.0, 2.0);
    glEnd();

    // Draw grid lines
    glColor4f(0.2f, 0.2f, 0.2f, 0.6f);
    glLineWidth(0.7f);
    
    // Draw grid lines parallel to x-axis
    const float gridStep = 0.1f;
    for (float y = 0.0f; y <= 1.0f + gridStep/2.0f; y += gridStep) {
        glBegin(GL_LINE_STRIP);
        for (float x = 0.0f; x <= 2.0f + gridStep/2.0f; x += gridStep) {
            int i = static_cast<int>(std::min(x * Nx / 2, (float)(Nx-1)));
            int j = static_cast<int>(std::min(y * Ny, (float)(Ny-1)));
            float height = u_current[i][j][Nz/2];
            glVertex3f(std::min(x, 2.0f), std::min(y, 1.0f), height);
        }
        glEnd();
    }
    
    // Draw grid lines parallel to y-axis
    for (float x = 0.0f; x <= 2.0f + gridStep/2.0f; x += gridStep) {
        glBegin(GL_LINE_STRIP);
        for (float y = 0.0f; y <= 1.0f + gridStep/2.0f; y += gridStep) {
            int i = static_cast<int>(std::min(x * Nx / 2, (float)(Nx-1)));
            int j = static_cast<int>(std::min(y * Ny, (float)(Ny-1)));
            float height = u_current[i][j][Nz/2];
            glVertex3f(std::min(x, 2.0f), std::min(y, 1.0f), height);
        }
        glEnd();
    }

    // Draw wave surface with enhanced color gradient
    glLineWidth(1.0f);
    for (int i = 0; i < Nx-1; i++) {
        glBegin(GL_TRIANGLE_STRIP);
        for (int j = 0; j < Ny; j++) {
            float height1 = u_current[i][j][Nz/2];
            float height2 = u_current[i+1][j][Nz/2];
            
            // Get colors for both vertices
            Color c1 = getHeightColor(height1);
            Color c2 = getHeightColor(height2);
            
            // First vertex
            glColor4f(c1.r, c1.g, c1.b, 0.75f);
            glVertex3f(i*dx, j*dy, height1);
            
            // Second vertex
            glColor4f(c2.r, c2.g, c2.b, 0.75f);
            glVertex3f((i+1)*dx, j*dy, height2);
        }
        glEnd();
    }

    // Draw contour lines with darker coloring
    glLineWidth(1.5f);  // Made lines slightly thicker
    
    const float contourInterval = 0.05f;
    for (int i = 0; i < Nx-1; i++) {
        for (int j = 0; j < Ny-1; j++) {
            float h00 = u_current[i][j][Nz/2];
            float h10 = u_current[i+1][j][Nz/2];
            float h01 = u_current[i][j+1][Nz/2];
            
            for (float level = -1.0f; level <= 1.0f; level += contourInterval) {
                // Get base color and darken it
                Color levelColor = getHeightColor(level);
                float darkFactor = 0.5f;  // Make contour lines darker
                glColor4f(
                    levelColor.r * darkFactor, 
                    levelColor.g * darkFactor, 
                    levelColor.b * darkFactor, 
                    0.9f  // More opaque
                );
                
                if ((h00 <= level && h10 >= level) || 
                    (h00 >= level && h10 <= level)) {
                    float t = (level - h00) / (h10 - h00);
                    glBegin(GL_LINES);
                    glVertex3f(i*dx + t*dx, j*dy, level);
                    glVertex3f(i*dx + t*dx, (j+1)*dy, level);
                    glEnd();
                }
                
                if ((h00 <= level && h01 >= level) || 
                    (h00 >= level && h01 <= level)) {
                    float t = (level - h00) / (h01 - h00);
                    glBegin(GL_LINES);
                    glVertex3f(i*dx, j*dy + t*dy, level);
                    glVertex3f((i+1)*dx, j*dy + t*dy, level);
                    glEnd();
                }
            }
        }
    }

    // Draw the pillar
    drawPillar();
    
    glutSwapBuffers();

    if (simulationRunning) {
        calculateNextStep();
    }

    std::this_thread::sleep_for(std::chrono::milliseconds(16));
    glutPostRedisplay();
}

int main(int argc, char** argv) {
    // Check CFL condition
    double maxC = calculateMaxStableC();
    std::cout << "Maximum stable wave speed (c): " << maxC << std::endl;
    if (c > maxC) {
        std::cerr << "Warning: Current c (" << c << ") exceeds maximum stable value!" << std::endl;
        std::cerr << "Simulation may become unstable." << std::endl;
    }

    // Initialize wave field with Gaussian pulse
    for (int i = 0; i < Nx; i++) {
        for (int j = 0; j < Ny; j++) {
            for (int k = 0; k < Nz; k++) {
                double x = i * dx;
                double y = j * dy;
                double z = k * dz;
                u_prev[i][j][k] = std::exp(-100.0 * (
                    (x - 0.5)*(x - 0.5) +
                    (y - 0.5)*(y - 0.5) + 
                    (z - 1.0)*(z - 1.0)));
                u_current[i][j][k] = u_prev[i][j][k];

                // Set initial conditions to zero inside the pillar
                if (i >= pillar_start_x && i < pillar_end_x &&
                    j >= pillar_start_y && j < pillar_end_y &&
                    k >= pillar_start_z && k < pillar_end_z) {
                    u_prev[i][j][k] = 0;
                    u_current[i][j][k] = 0;
                }
            }
        }
    }

    // Initialize OpenGL
    glutInit(&argc, argv);
    glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB | GLUT_DEPTH);
    glutInitWindowSize(1200, 600);  // Wider window for elongated domain
    glutCreateWindow("Wave Simulation with Obstacle");
    init();
    glutDisplayFunc(display);
    glutMainLoop();

    return 0;
}